dbt_style_guide.md

```diff
@@ -1,237 +1,415 @@
     drivers.rating as driver_rating,
     riders.rating as rider_rating
-
-from trips
-left join users as drivers
+  from trips
+  left join users as drivers
     on trips.driver_id = drivers.user_id
-left join users as riders
+  left join users as riders
     on trips.rider_id = riders.user_id
 
-```
+  ```
+
+- ### Example SQL
+  ```sql
+  with
+
+  my_data as (
+      select * from {{ ref('my_data') }}
+      where not is_deleted
+  ),
+
+  some_cte as (
+      select * from {{ ref('some_cte') }}
+  ),
+
+  some_cte_agg as (
+      select
+          id,
+          sum(field_4) as total_field_4,
+          max(field_5) as max_field_5
+      from some_cte
+      group by 1
+  ),
+
+  final as (
+      select [distinct]
+          my_data.field_1,
+          my_data.field_2,
+          my_data.field_3,
+
+          -- use line breaks to visually separate calculations into blocks
+          case
+              when my_data.cancellation_date is null
+                and my_data.expiration_date is not null
+                then expiration_data
+              when my_data.cancellation_date is null
+                then my_data.start_date + 7
+              else my_data.cancellation_date
+          end as cancellation_date,
+
+          some_cte_agg.total_field_4,
+          some_cte_agg.max_field_5
+      from my_data
+      left join some_cte_agg  
+          on my_data.id = some_cte_agg.id
+      where 
+        my_data.field_1 = 'abc'
+        and (
+            my_data.field_2 = 'def'
+            or my_data.field_2 = 'ghi'
+        )
+      having count(*) > 1
+  )
+
+  select * from final
+  ```
 
 ## YAML style guide
 
-* Indents should be two spaces
-* List items should be indented
-* Use a new line to separate list items that are dictionaries where appropriate
-* Lines of YAML should be no longer than 80 characters.
-
-### Example YAML
-```yaml
-version: 2
-
-models:
-  - name: events
-    columns:
-      - name: event_id
-        description: This is a unique identifier for the event
-        tests:
-          - unique
-          - not_null
-
-      - name: event_time
-        description: "When the event occurred in UTC (eg. 2018-01-01 12:00:00)"
-        tests:
-          - not_null
-
-      - name: user_id
-        description: The ID of the user who recorded the event
-        tests:
-          - not_null
-          - relationships:
-              to: ref('users')
-              field: id
-```
+- Every subdirectory contains their own `.yml` file(s) which contain configurations for the models within the subdirectory.
+
+- A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 
+
+- YAML files should be prefixed with an underscore to keep it at the top of the subdirectory.  
+  Example: `__sources_core.yml`, `_models_core.yml`
+
+- YAML files should be named for the top-level configuration it contains.  
+  Examples: `_sources_core.yml`, `_models_core.yml`

```

Comment:
```suggestion
- YAML files should be named for the top-level configuration it contains.  
  Examples: `_jaffle_shop__models.yml`, `_jaffle_shop__sources.yml`

```

Reply:
^That is not consistent with the prior examples.

Reply:
Generally, names should go `general_to_specific` so that they alphabetize correctly. 
You want a list like:
```
core_models
core_sources
customers
customers__pivoted
orders
orders__pivoted
```
not like:
```
customers
models_core
orders
pivoted__customers
pivoted__orders
sources_core
```

Reply:
similar to date formats
good:
```
2019-11-23
2020-04-12
2022-08-18
```
bad
```
04-12-2020
08-18-2022
11-23-2019
```

Reply:
Again I think we voted on having `_<config>_<thing>.yml` early on (possibly in Slack? Or in previous comments here?) Maybe we should all vote on this again. For config files I personally prefer the specific > general for this use case. For example, I wouldn't want `customers__jaffle_shop_stg.sql` as my model file name.

Reply:
Although, I guess it depends on how you look at it and how you're structuring it. If I put all my yml configs / .mds for my models at the staging level but had subfolders as we typically do, then the more general would be the _docs, _sources, or _models and the specific would be the folder:
```
_docs_asana.md
_docs_jaffle_shop.md
_docs_snowplow.md
_source_asana.yml
_source_jaffle_shop.yml
_source_snowplow.yml
```

Reply:
Are you disagreeing with yourself with this example?
> For example, I wouldn't want customers__jaffle_shop_stg.sql as my model file name.
customers is specific, it is the name of the table, which is more specific than the schema
jaffle_shop is the less specific, it is the bucket which holds customers
stg_ is general ie. is it staging or marts, and a folder up.

the way we write it is most general to most specific: `stg_jaffle_shop__customers`

Reply:
>then the more general would be the _docs, _sources, or _models and the specific would be the folder:

I think you're thinking about this backwards. 🤔

Reply:
my preference is `_directory_models.yml`, `_directory_sources.yml `, `_directory_docs.md`, `_directory_exposures.yml`. In theory you should only have one of each of these in each subdirectory so there shouldn't be any issues with ordering.

Reply:
^ A reminder that this vote is only for .yml files and that there can be different conventions for .yml files as opposed to models. Also that the underscore in front is vital to proper file structure.

Reply:
I agree with Sam and Pat here. I think the layout Sam provided in his example above is the correct generic > specific setup. This also aligns with the current guide on ["How we structure our dbt projects"](https://docs.getdbt.com/guides/best-practices/how-we-structure/1-guide-overview#:~:text=_jaffle_shop__models.yml)

Reply:
Changed the conventions to use what's in the article and additionally changed it a bit to include .md in this section, since most of the rules apply. View it [here](https://github.com/dbt-labs/corp/pull/63/commits/42e7f44c0e857f87286595855fc8b3518e9a8944)!